File: tests/test_physics.py
"""Physics unit tests against the wasmtime-hosted core.

Everything here is a property the design note states outright, checked
against the real wasm module rather than a Python model of it.
"""

from __future__ import annotations

import math

from conftest import require_sim_wasm

from cogball import defaults
from cogball.sim import EV_GOAL, EV_KICK, EV_POST, CogballSim

NO_CONTROLS = bytes(defaults.NUM_ROBOTS * 3)


def fresh(seed: int = 1, kickoff_seat: int = 0) -> CogballSim:
    require_sim_wasm()
    return CogballSim(seed=seed, first_kickoff_seat=kickoff_seat)


def park_robots(sim: CogballSim, x: float = -19.0, y: float = 11.0) -> None:
    """Stack every robot in a corner, out of the way of a ball experiment."""
    for i in range(defaults.NUM_ROBOTS):
        sim.debug_place_robot(i, x + 1.2 * i, y)


def ball(sim: CogballSim):
    state = sim.state()
    return state[0], state[1], state[2], state[3]


def robot(sim: CogballSim, i: int):
    base = 4 + i * 8
    s = sim.state()
    return s[base:base + 8]


def test_ball_fired_at_a_wall_never_tunnels():
    """30 m/s is 0.25 m per substep, comfortably inside the ball's radius."""
    sim = fresh()
    park_robots(sim)
    sim.debug_place_ball(0.0, 8.0, 30.0, 0.0)
    for _ in range(600):
        sim.set_controls(NO_CONTROLS)
        sim.step()
        bx, by, _, _ = ball(sim)
        assert abs(bx) <= defaults.PITCH_X + 2.0 + 1e-9
        assert abs(by) <= defaults.PITCH_Y + 1e-9
    assert not sim.fault()


def test_wall_bounce_reproduces_the_restitution():
    sim = fresh()
    park_robots(sim)
    sim.debug_place_ball(19.0, 8.0, 10.0, 0.0)
    for _ in range(30):
        sim.set_controls(NO_CONTROLS)
        sim.step()
        _, _, vx, _ = ball(sim)
        if vx < 0:
            break
    _, _, vx, _ = ball(sim)
    # restitution 0.80, minus a substep or two of the 0.6/s linear damping
    assert -8.1 < vx < -7.5, vx


def test_robot_robot_resolution_is_symmetric():
    """Mirror-image robots must produce a mirror-image outcome exactly."""
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    sim.debug_place_ball(0.0, -11.9, 0.0, 0.0)
    sim.debug_place_robot(0, -0.5, 6.0, 1.0, 0.0, 3.0, 0.0)
    sim.debug_place_robot(1, 0.5, 6.0, -1.0, 0.0, -3.0, 0.0)
    sim.set_controls(NO_CONTROLS)
    sim.step()
    a = robot(sim, 0)
    b = robot(sim, 1)
    assert a[0] == -b[0]          # positions mirror exactly
    assert a[1] == b[1]
    assert a[2] == -b[2]          # velocities mirror exactly
    assert a[3] == b[3]
    # equal masses, equal-and-opposite impulse: net x momentum stays zero
    assert abs(a[2] + b[2]) < 1e-12


def test_kick_sets_the_along_heading_speed_and_recoils_the_robot():
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    sim.debug_place_robot(0, 0.0, 0.0, 1.0, 0.0)
    sim.debug_place_ball(0.9, 0.0, 0.0, 0.0)
    ctl = bytearray(NO_CONTROLS)
    ctl[2] = 1
    sim.set_controls(bytes(ctl))
    sim.step()
    kicks = [e for e in sim.events() if int(e[0]) == EV_KICK]
    assert len(kicks) == 1
    # max(v_parallel, 0) + 9.0, exactly, along the heading
    assert kicks[0][6] == 9.0
    assert kicks[0][7] == 0.0
    # reaction: 0.45 * 9.0 / 6.0 = 0.675 m/s backwards (then substepped)
    assert robot(sim, 0)[2] < 0.0


def test_kick_respects_range_facing_and_cooldown():
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    ctl = bytearray(NO_CONTROLS)
    ctl[2] = 1

    # out of range
    sim.debug_place_robot(0, 0.0, 0.0, 1.0, 0.0)
    sim.debug_place_ball(2.0, 0.0, 0.0, 0.0)
    sim.set_controls(bytes(ctl))
    sim.step()
    assert not [e for e in sim.events() if int(e[0]) == EV_KICK]

    # in range but facing away
    sim.debug_place_robot(0, 0.0, 0.0, -1.0, 0.0)
    sim.debug_place_ball(1.0, 0.0, 0.0, 0.0)
    sim.set_controls(bytes(ctl))
    sim.step()
    assert not [e for e in sim.events() if int(e[0]) == EV_KICK]

    # in range and facing: fires, then the cooldown blocks the next tick
    sim.debug_place_robot(0, 0.0, 0.0, 1.0, 0.0)
    sim.debug_place_ball(1.0, 0.0, 0.0, 0.0)
    sim.set_controls(bytes(ctl))
    sim.step()
    assert [e for e in sim.events() if int(e[0]) == EV_KICK]
    sim.debug_place_ball(1.0, 0.0, 0.0, 0.0)
    sim.set_controls(bytes(ctl))
    sim.step()
    assert not [e for e in sim.events() if int(e[0]) == EV_KICK]


def test_goal_fires_on_the_plane_crossing_not_a_tick_early_or_late():
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    sim.debug_place_ball(18.0, 0.0, 8.0, 0.0)
    goal = None
    previous_x = 18.0
    for _ in range(120):
        previous_x = ball(sim)[0]
        sim.set_controls(NO_CONTROLS)
        sim.step()
        found = [e for e in sim.events() if int(e[0]) == EV_GOAL]
        if found:
            goal = found[0]
            break
    assert goal is not None, "the ball never reached the goal"
    assert previous_x < defaults.PITCH_X          # not one tick early
    assert goal[7] >= defaults.PITCH_X            # crossed the plane
    assert abs(goal[8]) <= defaults.GOAL_HALF_WIDTH
    assert int(goal[2]) == 0                      # Azure attacks +x
    assert sim.goals(0) == 1


def test_ball_on_a_post_bounces_and_reports_it():
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    sim.debug_place_ball(18.5, defaults.GOAL_HALF_WIDTH, 6.0, 0.0)
    hit = None
    for _ in range(60):
        sim.set_controls(NO_CONTROLS)
        sim.step()
        found = [e for e in sim.events() if int(e[0]) == EV_POST]
        if found:
            hit = found[0]
            break
    assert hit is not None, "the ball never touched the post"
    assert abs(hit[4]) == defaults.PITCH_X
    assert abs(hit[5]) == defaults.GOAL_HALF_WIDTH
    assert ball(sim)[2] < 6.0     # the post took pace off it


def test_kickoff_places_all_seven_bodies_where_the_note_says():
    for kickoff_seat in (0, 1):
        sim = fresh(seed=42, kickoff_seat=kickoff_seat)
        state = sim.state()
        assert (state[0], state[1], state[2], state[3]) == (0.0, 0.0, 0.0, 0.0)
        own = -1.0 if kickoff_seat == 0 else 1.0
        other_base = (1 - kickoff_seat) * defaults.ROBOTS_PER_SEAT
        base = kickoff_seat * defaults.ROBOTS_PER_SEAT
        assert robot(sim, base)[0] == own * 1.5
        assert robot(sim, base)[1] == 0.0
        assert robot(sim, other_base)[0] == -own * 3.0
        assert robot(sim, other_base)[1] == 0.0
        for slot, sign in ((1, 1.0), (2, -1.0)):
            wide = robot(sim, base + slot)
            assert wide[0] == own * 9.0
            assert abs(wide[1] - sign * 4.5) <= 0.25
            wide = robot(sim, other_base + slot)
            assert wide[0] == -own * 9.0
            assert abs(wide[1] - sign * 4.5) <= 0.25
        for i in range(defaults.NUM_ROBOTS):
            r = robot(sim, i)
            assert (r[2], r[3], r[6], r[7]) == (0.0, 0.0, 0.0, 0.0)
            expect = 1.0 if defaults.seat_of_robot(i) == 0 else -1.0
            assert (r[4], r[5]) == (expect, 0.0)


def test_a_goal_freezes_play_for_one_second_then_restarts():
    sim = fresh()
    park_robots(sim, x=-19.0, y=-11.0)
    sim.debug_place_ball(19.5, 0.0, 8.0, 0.0)
    while not sim.goals(0):
        sim.set_controls(NO_CONTROLS)
        sim.step()
    assert sim.frozen()
    frozen_ticks = 0
    while sim.frozen():
        sim.set_controls(NO_CONTROLS)
        sim.step()
        frozen_ticks += 1
    assert frozen_ticks == 30      # 1.0 s of kickoff freeze
    bx, by, bvx, bvy = ball(sim)
    assert (bx, by, bvx, bvy) == (0.0, 0.0, 0.0, 0.0)


def test_robots_stay_inside_the_pitch_and_the_ball_stays_in_the_arena():
    """Random-ish flailing for a while must never leave the world."""
    sim = fresh(seed=7)
    for tick in range(1200):
        ctl = bytearray(defaults.NUM_ROBOTS * 3)
        for i in range(defaults.NUM_ROBOTS):
            ctl[i * 3 + 0] = (100 - (tick * 37 + i * 11) % 200) & 0xFF
            ctl[i * 3 + 1] = (100 - (tick * 53 + i * 17) % 200) & 0xFF
            ctl[i * 3 + 2] = 1 if (tick + i) % 5 == 0 else 0
        sim.set_controls(bytes(ctl))
        sim.step()
        state = sim.state()
        assert abs(state[0]) <= 22.0 + 1e-9
        assert abs(state[1]) <= defaults.PITCH_Y + 1e-9
        for i in range(defaults.NUM_ROBOTS):
            r = state[4 + i * 8: 12 + i * 8]
            assert abs(r[0]) <= defaults.PITCH_X - defaults.ROBOT_R + 1e-9
            assert abs(r[1]) <= defaults.PITCH_Y - defaults.ROBOT_R + 1e-9
            assert math.isclose(math.hypot(r[4], r[5]), 1.0, rel_tol=1e-12)
            assert math.hypot(r[2], r[3]) <= defaults.ROBOT_MAX_SPEED + 1e-9
    assert not sim.fault()

File: tests/test_physics.nim
## Sim unit tests: the integer physics, the kick model, the restarts.
##
## These are the assertions that keep `sim.nim` honest about the things the
## design pins numerically. A failure here is a rules change, not a flake.

import std/math
import lib/helpers

proc noTunnelling() =
  ## A ball fired at BallMaxSpeed into a wall for 600 ticks never leaves the
  ## arena. Four substeps of 1/96 s move a 25 m/s ball at most 0.26 m — less
  ## than its own 0.70 m diameter — which is why nothing can pass through.
  for angle in 0 ..< 16:
    var sim = playing(testConfig())
    # Park every robot in a corner so only the ball is in play.
    for i in 0 ..< RobotCount:
      sim.robots[i].x = CentreX + int32((i - 3) * 300_000)
      sim.robots[i].y = 1_000_000'i32
      sim.robots[i].vx = 0
      sim.robots[i].vy = 0
    sim.ball.x = CentreX
    sim.ball.y = CentreY
    let b = int32(angle * 16)
    sim.ball.vx = int32((int64(BallMaxSpeed) * int64(cosQ12(b))) div 4096)
    sim.ball.vy = int32((int64(BallMaxSpeed) * int64(-sinQ12(b))) div 4096)
    for _ in 0 ..< 600:
      if sim.phase != Playing:
        break
      sim.stepIdle()
      doAssert inPitch(sim.ball.x, sim.ball.y),
        "ball left the arena at angle " & $angle & ": " &
          $sim.ball.x & "," & $sim.ball.y
      doAssert sim.endRule != erSimFault, "physics guard tripped"
  report "a max-speed ball never tunnels out of the arena"

proc wallRestitution() =
  ## The touchline reflects the normal component at 80 % and shaves the
  ## tangential at 98 %, within the fixed-point quantum.
  var sim = playing(testConfig())
  for i in 0 ..< RobotCount:
    sim.robots[i].x = CentreX
    sim.robots[i].y = CentreY + 8_000_000'i32
  sim.ball.x = CentreX
  sim.ball.y = 500_000'i32
  sim.ball.vx = 0
  sim.ball.vy = -400_000'i32
  let before = -sim.ball.vy
  sim.stepIdle()
  doAssert sim.ball.vy > 0, "the ball did not bounce off the touchline"
  # One integration substep of drag runs before the bounce, so allow for it.
  let expected = int32((int64(before) * 80) div 100)
  doAssert abs(sim.ball.vy - expected) <= expected div 8,
    "wall bounce speed " & $sim.ball.vy & " is nowhere near " & $expected
  report "wall restitution reproduces the analytic bounce"

proc robotPairSymmetry() =
  ## Robot-robot resolution is symmetric under swapping the indices and
  ## conserves momentum EXACTLY (Nim's div truncates toward zero, so the
  ## impulse is equal and opposite by construction).
  var sim = playing(testConfig())
  for i in 0 ..< RobotCount:
    sim.robots[i].x = CentreX + int32(i * 8_000_000) div 4
    sim.robots[i].y = 2_000_000'i32
    sim.robots[i].vx = 0
    sim.robots[i].vy = 0
  sim.ball.x = 20_000_000'i32
  sim.ball.y = 23_000_000'i32
  sim.robots[0].x = CentreX - 500_000'i32
  sim.robots[0].y = CentreY
  sim.robots[0].vx = 200_000'i32
  sim.robots[1].x = CentreX + 500_000'i32
  sim.robots[1].y = CentreY
  sim.robots[1].vx = -200_000'i32
  let sumBefore = int64(sim.robots[0].vx) + int64(sim.robots[1].vx)
  sim.stepIdle()
  let sumAfter = int64(sim.robots[0].vx) + int64(sim.robots[1].vx)
  doAssert sumBefore == 0
  doAssert sumAfter == 0, "momentum was not conserved: " & $sumAfter
  doAssert sim.robots[0].vx < 0 and sim.robots[1].vx > 0,
    "the pair did not separate"
  doAssert sim.robots[0].vx == -sim.robots[1].vx,
    "the resolution is not symmetric under index swap"
  report "robot-robot resolution is symmetric and conserves momentum"

proc kickModel() =
  ## The kick sets the along-heading ball speed to exactly
  ## max(vpar, 0) + KickImpulse and applies the mass-ratio reaction.
  var sim = playing(testConfig())
  for i in 0 ..< RobotCount:
    sim.robots[i].x = 6_000_000'i32 + int32(i) * 400_000'i32
    sim.robots[i].y = 23_000_000'i32
    sim.robots[i].vx = 0
    sim.robots[i].vy = 0
    sim.robots[i].headingQ = 0
    sim.robots[i].kickCooldown = 1
  sim.robots[0].x = CentreX
  sim.robots[0].y = CentreY
  sim.robots[0].kickCooldown = 0
  sim.ball.x = CentreX + 1_000_000'i32
  sim.ball.y = CentreY
  sim.ball.vx = 0
  sim.ball.vy = 0
  var masks: array[RobotCount, uint8]
  masks[0] = ButtonA
  sim.stepWith(masks)
  doAssert sim.robots[0].kickCooldown == KickCooldownTicks - 1,
    "the cooldown was not armed: " & $sim.robots[0].kickCooldown
  # One tick of drag and one substep of ball integration follow the kick.
  doAssert sim.ball.vx > int32(KickImpulse) * 9 div 10,
    "the kick did not reach the impulse: " & $sim.ball.vx
  doAssert sim.robots[0].vx < 0, "the robot took no reaction"
  let expectedReaction = int32(
    (int64(BallMassG) * int64(KickImpulse)) div int64(RobotMassG))
  doAssert abs(abs(sim.robots[0].vx) - expectedReaction) <=
      expectedReaction div 3,
    "the reaction is not the mass ratio: " & $sim.robots[0].vx &
      " vs " & $expectedReaction
  report "the kick applies the exact impulse and the mass-ratio reaction"

proc kickArc() =
  ## The +-60 degree frontal arc is exact: a ball behind the robot is not
  ## kickable however close it is.
  var sim = playing(testConfig())
  for i in 0 ..< RobotCount:
    sim.robots[i].x = 6_000_000'i32 + int32(i) * 400_000'i32
    sim.robots[i].y = 23_000_000'i32
    sim.robots[i].kickCooldown = 1
  sim.robots[0].x = CentreX
  sim.robots[0].y = CentreY
  sim.robots[0].headingQ = 0                 ## facing east
  sim.robots[0].kickCooldown = 0
  sim.ball.x = CentreX - 1_000_000'i32       ## directly behind
  sim.ball.y = CentreY
  var masks: array[RobotCount, uint8]
  masks[0] = ButtonA
  sim.stepWith(masks)
  doAssert sim.robots[0].kickCooldown == 0,
    "a ball behind the robot was kicked"
  report "the frontal arc rejects a ball behind the robot"

proc goalPlane() =
  ## The goal test fires on the exact plane crossing, and not one tick early.
  var sim = playing(testConfig())
  for i in 0 ..< RobotCount:
    sim.robots[i].x = CentreX + int32(i) * 400_000'i32
    sim.robots[i].y = 23_000_000'i32
    sim.robots[i].vx = 0
    sim.robots[i].vy = 0
  sim.ball.x = PitchXMax - 150_000'i32
  sim.ball.y = CentreY
  sim.ball.vx = 400_000'i32
  sim.ball.vy = 0
  doAssert goalScoredBy(sim.ball.x, sim.ball.y) < 0
  doAssert sim.goals(Azure) == 0
  sim.stepIdle()
  doAssert sim.goals(Azure) == 1, "the goal did not fire on the crossing"
  doAssert sim.ball.x == CentreX and sim.ball.y == CentreY,
    "the kickoff reset did not place the ball on the centre spot"
  doAssert sim.freezeUntil > int32(sim.tickCount) - 1,
    "the kickoff freeze was not armed"
  # And not a millimetre early: a ball just short of the plane scores nothing.
  var short = playing(testConfig())
  short.ball.x = PitchXMax - 1'i32
  short.ball.y = CentreY
  doAssert goalScoredBy(short.ball.x, short.ball.y) < 0
  doAssert goalScoredBy(PitchXMax, CentreY) == int32(ord(Azure))
  doAssert goalScoredBy(PitchXMin, CentreY) == int32(ord(Crimson))
  doAssert goalScoredBy(PitchXMax, GoalYMin - 1) < 0,
    "a ball outside the mouth band scored"
  report "the goal test fires on the exact plane crossing"

proc postBounce() =
  ## A ball rolled onto a post bounces and emits `post`.
  var sim = playing(testConfig())
  sim.collectEvents = true
  for i in 0 ..< RobotCount:
    sim.robots[i].x = CentreX + int32(i) * 400_000'i32
    sim.robots[i].y = 23_000_000'i32
  sim.ball.x = PitchXMax - 1_200_000'i32
  sim.ball.y = GoalYMin
  sim.ball.vx = 300_000'i32
  sim.ball.vy = 0
  var sawPost = false
  for _ in 0 ..< 20:
    sim.stepIdle()
    for event in sim.events:
      if event.kind == Post:
        sawPost = true
    sim.events.setLen(0)
    if sawPost:
      break
  doAssert sawPost, "the ball did not touch a post"
  doAssert sim.ball.vx <= 0 or sim.ball.vy != 0,
    "the post did not deflect the ball"
  report "a ball on the post bounces and emits `post`"

proc kickoffPlacement() =
  ## The kickoff places all seven bodies at the documented coordinates, jitter
  ## included, and is seed-pinned.
  var sim = seatedSim(testConfig())
  sim.startGame()
  doAssert sim.ball.x == CentreX and sim.ball.y == CentreY
  doAssert sim.ball.vx == 0 and sim.ball.vy == 0
  let restart = Seat(sim.restartSeat and 1)
  for i in 0 ..< RobotCount:
    let
      seat = seatOfRobot(i)
      slot = i mod RobotsPerSeat
    doAssert sim.robots[i].vx == 0 and sim.robots[i].vy == 0
    doAssert sim.robots[i].spin == 0
    doAssert sim.robots[i].kickCooldown == 0
    doAssert sim.robots[i].headingQ ==
      (if seat == Azure: 0'i32 else: 2048'i32)
    if slot == 0:
      let want =
        if seat == restart: CentreX - 1_500_000'i32 * attackDir(seat)
        else: CentreX - 3_000_000'i32 * attackDir(seat)
      doAssert sim.robots[i].x == want,
        "robot " & robotId(i) & " x " & $sim.robots[i].x & " != " & $want
      doAssert sim.robots[i].y == CentreY
    else:
      doAssert sim.robots[i].x == CentreX - 9_000_000'i32 * attackDir(seat)
      let base = CentreY + (if slot == 1: -4_500_000'i32 else: 4_500_000'i32)
      doAssert abs(sim.robots[i].y - base) <= 250_000'i32,
        "flank jitter out of range: " & $(sim.robots[i].y - base)
  # Seed-pinned: the same seed produces byte-identical placement.
  var again = seatedSim(testConfig())
  again.startGame()
  for i in 0 ..< RobotCount:
    doAssert again.robots[i].y == sim.robots[i].y,
      "the kickoff jitter is not seed-pinned"
  report "the kickoff places all seven bodies exactly, jitter seed-pinned"

proc neutralDropFires() =
  ## The drop fires at exactly `stalemateTicks` and clears the 3 m radius.
  var config = testConfig()
  var sim = playing(config)
  sim.collectEvents = true
  # Park the ball and every robot dead still in a corner.
  sim.ball.x = 3_000_000'i32
  sim.ball.y = 1_000_000'i32
  sim.ball.vx = 0
  sim.ball.vy = 0
  sim.anchorX = sim.ball.x
  sim.anchorY = sim.ball.y
  sim.stalemateTicks = 0
  for i in 0 ..< RobotCount:
    sim.robots[i].x = 3_500_000'i32 + int32(i) * 200_000'i32
    sim.robots[i].y = 1_500_000'i32
    sim.robots[i].vx = 0
    sim.robots[i].vy = 0
  for tick in 1 ..< config.stalemateTicks:
    sim.stepIdle()
    doAssert sim.stalemateTicks == int32(tick),
      "the stalemate counter did not advance: " & $sim.stalemateTicks
  sim.stepIdle()
  doAssert sim.stalemateTicks == 0, "the drop did not reset the counter"
  let spot = nearestDropSpot(3_000_000'i32, 1_000_000'i32)
  doAssert sim.ball.x == spot.x and sim.ball.y == spot.y,
    "the ball was not dropped on the nearest spot"
  for i in 0 ..< RobotCount:
    doAssert distI(sim.robots[i].x - spot.x, sim.robots[i].y - spot.y) >=
      DropClearRadius - 2,
      "robot " & robotId(i) & " is still inside the drop radius"
  report "the neutral drop fires at 240 ticks and clears the radius"

proc kickRangeMatchesArt() =
  ## The kick radius and the drawn robot are two numbers in two modules and
  ## nothing structurally ties them (AGENTS.md). Assert the relationship.
  doAssert KickRange == RobotRadius + BallRadius + KickReach
  doAssert KickReach > 0 and KickReach < RobotRadius,
    "the kick slack must be a fraction of the hull, not another hull"
  report "the kick radius is derived from the two body radii"

when isMainModule:
  echo "test_physics"
  noTunnelling()
  wallRestitution()
  robotPairSymmetry()
  kickModel()
  kickArc()
  goalPlane()
  postBounce()
  kickoffPlacement()
  neutralDropFires()
  kickRangeMatchesArt()
  echo "test_physics: all good"
